### PR TITLE
fix: remove hardcode precision in attach funds

### PR DESCRIPTION
**Pull request by bkioshn**
## Describe your changes
- Remove hardcode precision in attach funds dropdown selection. 
- Delete microfy and demicofy

### CODE REVIEW

File: src/lib/app-provider/hooks/useAttachFunds.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  return useCallback(
+    (
+      attachFundsOption: AttachFundsType,
+      assetsJsonStr: string,
+      assetsSelect: Coin[]
+    ) => {
+      const assetsSelectWithPrecision = assetsSelect.map((coin) => {
+        return {
+          ...coin,
+          precision: getAssetInfo(coin.denom)?.precision ?? 1,
```
Comment (songwongtp) [inline src/lib/app-provider/hooks/useAttachFunds.ts:21]: let the argument type be `Option<number>`, then let `exponentify()` handle the default case

**Review by songwongtp — CHANGES_REQUESTED** [review]

File: src/lib/utils/funds.ts [hunk 2]
```diff
@@ -1,18 +1,29 @@
...
     aDenom.localeCompare(bDenom)
   );
 
-export const fabricateFunds = (assets: Coin[]): Coin[] =>
-  sortDenoms(assets.filter((asset) => Number(asset.amount) && asset.denom)).map(
-    (asset) => ({
-      ...asset,
-      amount: microfy(asset.amount as Token).toFixed(0),
-    })
+interface CoinWithPrecision {
```
Comment (songwongtp) [inline src/lib/utils/funds.ts:12]: CoinWithPrecision extends Coin

File: src/lib/app-provider/hooks/useAttachFunds.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+import { fabricateFunds, sortDenoms } from "lib/utils";
+
+export const useAttachFunds = () => {
+  const { assetInfos } = useAssetInfos();
+
+  return useCallback(
+    (
+      attachFundsOption: AttachFundsType,
+      assetsJsonStr: string,
+      assetsSelect: Coin[]
```
Comment (poomthiti) [inline src/lib/app-provider/hooks/useAttachFunds.ts:15]: I prefer naming this selectedAssets

File: src/lib/app-provider/hooks/useAttachFunds.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        };
+      });
+
+      switch (attachFundsOption) {
+        case AttachFundsType.ATTACH_FUNDS_SELECT:
+          return fabricateFunds(assetsSelectWithPrecision);
+        case AttachFundsType.ATTACH_FUNDS_JSON:
+          try {
+            if (JSON.parse(assetsJsonStr)) {
+              return sortDenoms(JSON.parse(assetsJsonStr) as Coin[]);
```
Comment (poomthiti) [inline src/lib/app-provider/hooks/useAttachFunds.ts:30]: remove `as`

File: src/lib/app-provider/hooks/useAttachFunds.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+          precision: assetInfos?.[coin.denom]?.precision,
+        };
+      });
+
+      switch (attachFundsOption) {
+        case AttachFundsType.ATTACH_FUNDS_SELECT:
+          return fabricateFunds(assetsSelectWithPrecision);
+        case AttachFundsType.ATTACH_FUNDS_JSON:
+          try {
+            if (JSON.parse(assetsJsonStr)) {
```
Comment (poomthiti) [inline src/lib/app-provider/hooks/useAttachFunds.ts:29]: remove if

File: src/lib/app-provider/hooks/useAttachFunds.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+              return sortDenoms(JSON.parse(assetsJsonStr) as Coin[]);
+            }
+          } catch {
+            return [];
+          }
+          return [];
```
Comment (poomthiti) [inline src/lib/app-provider/hooks/useAttachFunds.ts:35]: remove

File: src/lib/app-provider/hooks/useBaseApiRoute.ts [hunk 7]
```diff
@@ -10,6 +10,7 @@ export const useBaseApiRoute = (
     | "codes"
     | "accounts"
     | "rest"
+    | "feeTokens"
```
Comment (poomthiti) [inline src/lib/app-provider/hooks/useBaseApiRoute.ts:13]: fee_tokens

**Review by poomthiti — CHANGES_REQUESTED** [review]

File: src/lib/app-provider/hooks/useBaseApiRoute.ts (hunk 7 above)
Comment (evilpeach) [reply src/lib/app-provider/hooks/useBaseApiRoute.ts:13]: Perhaps we could consider using `native_tokens`.

File: src/lib/app-provider/hooks/useAttachFunds.ts (hunk 3 above)
Comment (bkioshn) [reply src/lib/app-provider/hooks/useAttachFunds.ts:15]: the assetsSelect represent the asset in the Select asset and fill amount section. selectedAssets might cause a confusion too

File: src/lib/services/assetService.ts [hunk 8]
```diff
@@ -26,3 +27,12 @@ export const useAssetInfos = (): {
     isLoading,
   };
 };
+
+export const useNativeTokensInfo = (): UseQueryResult<Option<AssetInfo[]>> => {
```
Comment (evilpeach) [inline src/lib/services/assetService.ts:31]: We just use
```tsx
```

**Review by songwongtp — CHANGES_REQUESTED** [review]

File: src/lib/services/faucetService.ts [hunk 9]
```diff
@@ -31,7 +32,7 @@ export const useFaucetInfo = (): UseQueryResult<FaucetInfo> => {
   };
 
   return useQuery({
-    queryKey: ["query", "faucet_info", faucet],
+    queryKey: ["query", "faucet_info", faucet, assetInfos],
```
[the remark this answers is not in the document]
Comment (bkioshn) [reply src/lib/services/faucetService.ts:35]: You mean passing `faucet` right?

**Review by evilpeach — APPROVED** [review]

**Review by songwongtp — APPROVED** [review]

File: src/lib/app-provider/hooks/useBaseApiRoute.ts [hunk 10]
```diff
@@ -42,6 +43,8 @@ export const useBaseApiRoute = (
       return `${api}/accounts/${chain}/${currentChainId}`;
     case "rest":
       return `${api}/rest/${chain}/${currentChainId}`;
+    case "native_tokens":
```
Comment (evilpeach) [inline src/lib/app-provider/hooks/useBaseApiRoute.ts:46]: change to `native_assets`

**Review by poomthiti — APPROVED** [review]

**Pull request merged**